File: schemas/analytics-metrics/convex/goals.ts
// goals: Conversion goals.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const goals = defineTable({
  name: v.string(),
  description: v.optional(v.string()),
  goalType: v.union(v.literal("event"), v.literal("page_view"), v.literal("custom")),
  eventTypeId: v.optional(v.id("event_types")),
  urlPattern: v.optional(v.string()),
  value: v.optional(v.number()),
  isActive: v.boolean(),
  createdBy: v.id("users"),
  updatedAt: v.number(),
})
  .index("by_goal_type", ["goalType"])
  .index("by_event_type_id", ["eventTypeId"])
  .index("by_is_active", ["isActive"])
  .index("by_created_by", ["createdBy"]);

File: schemas/analytics-metrics/spacetimedb/goals.rs
// goals: Conversion goals that track specific user actions or page views.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum GoalType {
    Event,    // type: String
    PageView,
    Custom,
}

#[spacetimedb::table(name = goals, public)]
pub struct Goal {
    #[primary_key]
    pub id: String, // UUID

    pub name: String,
    pub description: Option<String>,
    #[index(btree)]
    pub goal_type: GoalType,

    #[index(btree)]
    pub event_type_id: Option<String>, // UUID, FK → event_types.id (set null)

    pub url_pattern: Option<String>,
    pub value: Option<f64>,

    #[index(btree)]
    pub is_active: bool,

    #[index(btree)]
    pub created_by: String, // UUID, FK → users.id (restrict delete)

    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
